feat(appmesh): add ClientPolicy to VirtualNode, VirtualGateway and VirtualService

sshver opened the pull request:
Adds backend defaults to Virtual Node and Virtual Gateways.
Adds validation context for the backend defined on the Virtual Node.

Before merging (TODO):

- [ ] Update README

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 7ca64da277a2b01be7b602022d8aad05d2b25967
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4ef409db-6ea5-4766-aed9-215558a82ddc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,58 @@
+/**
+ * TLS Connections with downstream server will always be enforced if True
+ */
+export interface TLSClientPolicyOptions {
```
This should be called `TlsClientPolicyOptions`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,58 @@
+  /**
+   * TLS enforced if True.
+   *
+   * @default - True
```
```suggestion
   * @default true
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,58 @@
+  /**
+   * TLS enforced on these ports. If not specified it is enforced on all ports.
+   *
+   * @default - none
```
This comment does not match what's in the description above...

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,58 @@
+/**
+ * ACM Trust Properties
+ */
+export interface ACMTrustOptions {
```
```suggestion
export interface AcmTrustOptions {
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/validation-context.ts:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Defines the TLS validation context trust.
+ */
+export abstract class TLSValidationContext {
```
```suggestion
export abstract class TlsValidationContext {
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/validation-context.ts:
```diff
@@ -0,0 +1,91 @@
+import * as cdk from '@aws-cdk/core';
+import { CfnVirtualNode } from './appmesh.generated';
+import { ACMTrustOptions, FileTrustOptions } from './client-policy';
```
What's the point of the split into 2 files, `validation-context.ts` and `client-policy.ts`? It looks to me like they should be all in one file...

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/validation-context.ts:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Properties of TLS validation context
+ */
+export interface TLSValidationConfig {
```
I'd move this interface to above its first usage, otherwise the file reads a little wonky.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -60,6 +61,13 @@ export interface VirtualGatewayBaseProps {
+   *
+   * @default - No Config
+   */
+  readonly backendDefaults?: ClientPolicy;
```
Why this mismatch between the property name, and the type name? This reads a little surprising to me... Will `ClientPolicy` be used for something else other than `backendDefaults`?

sshver replied:
ClientPolicy is being used for [backends](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-virtualnode-virtualservicebackend.html) and [backend defaults](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-virtualnode-backenddefaults.html).

skinny85 replied:
That doesn't really answer my question 🙂.

It's very weird to have a property called `backendDefaults`, and its type being `ClientPolicy`.

I see `ClientPolicy` is also used in `VirtualNode`, where the property is called `backendDefaults` too, and in `VirtualService`, where it's called `clientPolicy`.

Can you explain why the weird naming?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -151,6 +159,7 @@ export class VirtualGateway extends VirtualGatewayBase {
   public readonly mesh: IMesh;
 
   protected readonly listeners = new Array<VirtualGatewayListenerConfig>();
+  private readonly backendDefaults: CfnVirtualGateway.VirtualGatewayBackendDefaultsProperty;
```
Why does this need to be a field? It's only set and read in the constructor of the class?

sshver replied:
I have updated the PR to use an array here. This gives me an option set `omitEmptyArray` true and this would remove backendDefault if not defined.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -166,13 +175,20 @@ export class VirtualGateway extends VirtualGatewayBase {
     }
 
+    if (props.backendDefaults) {
+      this.backendDefaults = this.addBackendDefaults(props.backendDefaults);
```
I don't think this is a good method name. `add` should only be used if there can be multiples of the thing added. Since this doesn't seem to be the case here, I would go for something like `setBackendDefaults`.

sshver replied:
Updated the function name to `setBackendDefaults`

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -186,6 +202,21 @@ export class VirtualGateway extends VirtualGatewayBase {
+      clientPolicy: {
+        tls: {
+          enforce: clientPolicy.tlsClientPolicy.enforce ?? true,
+          validation: clientPolicy.tlsClientPolicy.validation.bind(this).tlsValidation,
```
Why is the `tslValidation` set in `bind()`, but the rest are properties of the `ClientPolicy`? Wondering why the different approach.

sshver replied:
Earlier, I was trying to separate out the validation context as it could be used in future for Listener TLS.
But now I have a single file `client-policy.ts` and now the `clientPolicy` properties are being set in `bind()`.

This would make the API more intuitive.

```typescript
new appmesh.VirtualGateway(stack, 'virtual-gateway', {
 virtualGatewayName: 'virtual-gateway',
 mesh: mesh,
 backendDefaults: appmesh.ClientPolicy.fileTrust({
   certificateChain: ['path-to-certificate'],
   enforceTls: false,
 }),
});
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -231,11 +260,19 @@ export class VirtualNode extends VirtualNodeBase {
+            ports: virtualService.clientPolicy?.tlsClientPolicy.ports,
+          },
+        } : undefined,
       },
```
Exactly the same comments in this class, as in `VirtualGateway`.

In particular, it looks like a fair amount of duplication could be eliminated by moving some of these properties to `bind()`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:
```diff
@@ -92,6 +99,13 @@ const node3 = mesh.addVirtualNode('node3', {
   })],
+  backendDefaults: {
+    tlsClientPolicy: {
+      validation: appmesh.TLSValidationContext.fileTrust({
```
We're testing `fileTrust` twice here. Should we use a different static factory method for the second time?

skinny85 reviewed (changes requested):
Looks good @sshver ! Some initial comments / questions.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 01be91bb04134e2f974035a0708c36f3f1b2e8c5
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d8512aad-bdf5-40dc-9a0e-b460dbb2746d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 7e2ddbcbc7912200ddaf7a8ae986785f89a79860
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b560538e-69bb-4214-b291-4df6aa6267ca%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -186,6 +199,13 @@ export class VirtualGateway extends VirtualGatewayBase {
+  /**
+   * Adds Default Backend Configuration for virtual node to communicate with Virtual Services.
+   */
+  private addBackendDefaults(backendDefaults: ClientPolicy) {
```
I would get rid of this method, and turn the `backendDefaults` field to a local variable.

Also, there can only be a single `ClientPolicy` passed in `backendDefaults`. Why, since the L1 class takes in an array?

sshver replied:
Updated and created a local constructor variable. The backend default would accepts an object instead of an array.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -216,7 +230,12 @@ export class VirtualNode extends VirtualNodeBase {
+  /**
+   * Adds Default Backend Configuration for virtual node to communicate with Virtual Services.
+   */
+  private setBackendDefaults(backendDefaults: ClientPolicy) {
```
Same comment here as above:

1. Get rid of the `backendDefaults` field and the `setBackendDefaults` method, just make it a local variable of the constructor.
2. Why does `backendDefaults` accept a single `ClientPolicy`, while the L1 class accepts an array?

sshver replied:
I have created `backendDefaults` as a local variable of the constructor. The backendDefaults accepts an a single `ClientPolicy` instead of an array. ([Reference](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-virtualnode-backenddefaults.html))

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -231,11 +250,13 @@ export class VirtualNode extends VirtualNodeBase {
   }
 }
 
+
```
Unnecessary empty line.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -27,6 +28,11 @@ export interface IVirtualService extends cdk.IResource {
+  /**
+   * Client policy for Virtual Service
+   */
+  readonly clientPolicy?: ClientPolicy;
```
It's interesting to add this here. Should it also be added to `VirtualServiceAttributes`, so it can be set for imported `VirtualService`s too?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -119,6 +132,11 @@ export class VirtualService extends cdk.Resource implements IVirtualService {
    */
   public readonly mesh: IMesh;
 
+  /**
+   * Client policy for Virtual Service
+   */
```
You don't need the second comment if the property is coming from an interface.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:
```diff
@@ -96,6 +100,10 @@ const node3 = mesh.addVirtualNode('node3', {
       unhealthyThreshold: 2,
     },
   })],
+  backendDefaults: appmesh.ClientPolicy.fileTrust({
```
Can we test the other static factory method here, `acmTrust`?

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -251,7 +250,139 @@ export = {
+        // THEN
+        expect(stack).to(haveResourceLike('AWS::AppMesh::VirtualNode', {
+          Spec: {
+            Listeners: [
+              {
+                HealthCheck: {
+                  HealthyThreshold: 2,
+                  IntervalMillis: 5000,
+                  Port: 80,
+                  Protocol: 'tcp',
+                  TimeoutMillis: 3000,
+                  UnhealthyThreshold: 2,
+                },
+                PortMapping: {
+                  Port: 80,
+                  Protocol: 'tcp',
+                },
+                Timeout: {
+                  TCP: {
+                    Idle: {
+                      Unit: 'ms',
+                      Value: 10000,
+                    },
+                  },
+                },
+              },
+            ],
```
Is there a reason this is included in this test? I think we're only interested in the below `Backends` part, right?

sshver replied:
Removed Listeners and kept backend part in test.

skinny85 reviewed (changes requested):
Looks good @sshver , but I still have some questions 🙂

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 0306a36ea49b26942be49dac7447f2eb86ad8c89
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b218d218-885b-4a6a-b32f-9600b7f7d0c9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

sshver commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,108 @@
+  /**
+   * Amazon Resource Names (ARN) of trusted ACM Private Certificate Authorities
+   */
+  readonly certificateAuthorityArns: string[];
```
Added a skeleton L2 for ACMPCA.

sshver commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,108 @@
+  /**
+   * Path to the Certificate Chain file on the file system where the Envoy is deployed.
+   */
+  readonly certificateChain: string[];
```
Updated to `string`

sshver commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,108 @@
+}
+
+class ClientPolicyImpl extends ClientPolicy {
+  constructor (private readonly enforce: boolean = true,
```
Removed the default arguments

sshver commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,108 @@
+    };
+  }
+
+  private renderValidation(): CfnVirtualNode.TlsValidationContextProperty {
```
Removed the method `renderValidation` and added inline.

sshver commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,108 @@
+class ClientPolicyImpl extends ClientPolicy {
+  constructor (private readonly enforce: boolean = true,
+    private readonly ports: number[] | undefined,
+    private readonly certificateType: string,
```
created a enum for Certificate Type.

sshver commented:
In reply to [comment](https://github.com/aws/aws-cdk/pull/11563#discussion_r529776236),
I have created a shared interface `BackendDefaults` with `clientPolicy` as a property. The interface is used in `VirtualNode` and `VirtualGateway` to represent the backendDefault property. The `ClientPolicy` is a part of `VirtualServiceBackend` object. ([Reference](https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_VirtualServiceBackend.html)). Therefore I have updated the VirtualService and added a property `clientPolicy` of type `ClientPolicy`

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -187,14 +194,17 @@ export class VirtualNode extends VirtualNodeBase {
 
     props.backends?.forEach(backend => this.addBackend(backend));
     props.listeners?.forEach(listener => this.addListener(listener));
+    const backendDefaults = props.backendDefaults?.clientPolicy.bind(this);
```
Same here - I would inline this `backendDefaults` local variable.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -57,6 +63,13 @@ export interface VirtualServiceBaseProps {
   readonly virtualNode?: IVirtualNode;
+
+  /**
+   * Client policy for Virtual Service
```
```suggestion
   * Client policy for this Virtual Service.
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/package.json:
```diff
@@ -97,7 +98,8 @@
     "@aws-cdk/core": "0.0.0",
-    "constructs": "^3.2.0"
+    "constructs": "^3.2.0",
+    "@aws-cdk/aws-acmpca": "0.0.0"
```
Please add this so the alphabetical order of the dependencies is preserved, like you did for `"dependencies"`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:
```diff
@@ -62,6 +63,8 @@ router.addRoute('route-1', {
   }),
 });
 
+const arn = 'arn:aws:acm-pca:us-east-1:123456789012:certificate-authority/12345678-1234-1234-1234-123456789012';
```
Rename to `certificateAuthorityArn`?

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:
```diff
@@ -1,3 +1,4 @@
+import { CertificateAuthority } from '@aws-cdk/aws-acmpca';
```
Same here - no unqualified imports outside your module.

```suggestion
import * as acmpca from '@aws-cdk/aws-acmpca';
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:
```diff
@@ -1,3 +1,4 @@
+import { CertificateAuthority } from '@aws-cdk/aws-acmpca';
 import * as ec2 from '@aws-cdk/aws-ec2';
 import * as cloudmap from '@aws-cdk/aws-servicediscovery';
 import * as cdk from '@aws-cdk/core';
```
Kill the empty line below this line too.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -1,7 +1,7 @@
 import { expect, haveResourceLike } from '@aws-cdk/assert';
+import { CertificateAuthority } from '@aws-cdk/aws-acmpca';
```
Same here - no unqualified imports outside your module.

```suggestion
import * as acmpca from '@aws-cdk/aws-acmpca';
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -251,7 +251,112 @@ export = {
+                  Validation: {
+                    Trust: {
+                      ACM: {
+                        CertificateAuthorityArns: ['arn:aws:acm-pca:us-east-1:123456789012:certificate-authority/12345678-1234-1234-1234-123456789012'],
```
Use the variable from above here instead of repeating the value?

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -251,7 +251,112 @@ export = {
+            },
+          },
+        }));
+        test.done();
```
Empty line before this line.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -251,7 +251,112 @@ export = {
+            ],
+          },
+        }));
         test.done();
```
Empty line before this line.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -251,7 +251,112 @@ export = {
+          meshName: 'test-mesh',
+        });
+
+        const arn = 'arn:aws:acm-pca:us-east-1:123456789012:certificate-authority/12345678-1234-1234-1234-123456789012';
```
Rename to `certificateAuthorityArn`?

skinny85 reviewed (changes requested):
Looks great @sshver ! Nice work on adding the skeleton ACM PCA resources, I have a feeling that skeleton will get filled out fast after we merge this 😉.

A few minor comments before I give my "ShipIt".

skinny85 commented:
Also, the ReadMe changes are missing 🙂.

sshver commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,111 @@
+   *
+   * @default true
+   */
+  readonly enforceTls?: boolean;
```
I think I need to update the comments in code. If the `enforceTls` is `true` the virtual node communicates with all backends using Transport Layer Security (TLS). If we only want to require the use of TLS for one or more specific ports, then we specify them in `ports`. If we don't specify any ports, TLS is enforced for all ports that the `virtualnode` would use. And the default is `True`, if a value isn't specified.[Reference](https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_ClientPolicyTls.html)
So we would need `enforceTls` in case where we do not want to use TLS.

skinny85 replied:
So, to sum up:

1. `enforceTls: true` (or `undefined`, as that's the default), and `ports`: `undefined`:

TLS will be enforced for all ports.

2. `enforceTls: true` (or `undefined`, as that's the default), and `ports: [1234]`:

TLS will be enforced only for port 1234.

3. `enforceTls: false`, and `ports: undefined` (default):

TLS will not be enforced for any ports.

4. `enforceTls: false`, and `ports: [1234]`:

This seems like it's wrong, and should result in a validation error...?

sshver commented on packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:
```diff
@@ -138,4 +139,13 @@ class FileAccessLog extends AccessLog {
+  /**
+   * Represents Client Policy
+   */
+  readonly clientPolicy: ClientPolicy;
```
I am not entirely sure if the Backend Defaults would have more property in future.
In previous revision of the pull request, `backendDefaults` property was of type `ClientPolicy` in `VirtualNode` and `VirtualGateway` and we also had a `clientPolicy` property of type `ClientPolicy` in `VirtualService` which was weird.
So to fix this weird naming, I created a `BackendDefaults` interface. `backendDefaults` property is of type `BackendDefaults` in VirtualNode and VirtualGateway now.

skinny85 replied:
But that's a weird fix then 🙂.

Why not change the name of the properties in `VirtualNode` and `VirtualGateway` to something like `backendsDefaultClientPolicy`, and get rid of this artificial `BackendDefaults` interface?

sshver commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -165,14 +171,15 @@ export class VirtualGateway extends VirtualGatewayBase {
       meshName: this.mesh.meshName,
       spec: {
         listeners: this.listeners.map(listener => listener.listener),
+        backendDefaults: backendDefaults,
```
[ClientPolicy](https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_ClientPolicy.html) is different from a [VirtualGatewayBackendDefaults](https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_VirtualGatewayClientPolicy.html) i.e. they are two different datatype in AppMesh API reference, although structurally they are same. Many structure were duplicated to future proof in AppMesh.
Wouldn't it be okay to use any one of these, as structurally they are same (considering structural typing in Typescript) ?

skinny85 replied:
OK. If you are 100% certain `CfnVirtualNode.ClientPolicyProperty` and `CfnVirtualGateway.VirtualGatewayClientPolicyProperty` will never ever have different properties, you can 
leave this as now.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 2cad8826dc350079e09a4d0714ae62ca8830f766
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=171992bf-4622-4724-b3ac-91d2d62c5d44%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: da25261b17e522d07592d399e78636b9956234c1
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6ad13043-102a-43df-9f57-b06df62f0b88%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d18fdc4b96e8f88eaeb3c0d81f3922fa02657487
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=48d30157-d7b6-4698-b11b-7db250dfbd21%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -159,8 +159,22 @@ const node = mesh.addVirtualNode('virtual-node', {
+  }),
+});
+
+node.addBackend(virtualService);
```
This is a weird example, right? Because if I understand correctly, because the `VirtualService` defines its own `clientPolicy`, the `backendDefaultsClientPolicy` on `VirtualNode` will be ignored?

If my understanding is correct, I would re-structure the example to be less artificial.

skinny85 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -159,8 +159,22 @@ const node = mesh.addVirtualNode('virtual-node', {
     },
   })],
+  backendDefaultsClientPolicy: appmesh.ClientPolicy.fileTrust({
+    certificateChain: 'path-to-certificate',
```
Can we put a more realistic path to the certificate here? Where are they usually stored? Right now, this looks kind of weird as an example (it's not even a file path...).

skinny85 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -159,8 +159,22 @@ const node = mesh.addVirtualNode('virtual-node', {
+  virtualServiceName: 'service1.domain.local',
+  mesh,
+  clientPolicy: appmesh.ClientPolicy.fileTrust({
+    certificateChain: 'path-to-certificate',
```
Same comment for this path as above.

skinny85 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -184,14 +198,32 @@ const node = new VirtualNode(this, 'node', {
+  }),
+});
+
+node.addBackend(virtualService);
```
This example is exactly the same as above... what's the point of having the same identical example twice in the ReadMe?

skinny85 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -294,12 +329,17 @@ const gateway = mesh.addVirtualGateway('gateway', {
 The listeners field can be omitted which will default to an HTTP Listener on port 8080.
 A gateway route can be added using the `gateway.addGatewayRoute()` method.
 
+The `backendDefaults` property are added to the node while creating the virtual gateway. These are virtual gateway's service backends client policy defaults.
```
Old name used here (it should be `backendDefaultsClientPolicy`, not `backendDefaults`).

skinny85 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -294,12 +329,17 @@ const gateway = mesh.addVirtualGateway('gateway', {
         interval: cdk.Duration.seconds(10),
       },
   })],
+  backendDefaultsClientPolicy: appmesh.ClientPolicy.fileTrust({
```
We keep showing the `fileTrust()` Client Policy. Can we have an example of using the `acmTrust()` Client Policy too?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,110 @@
+
+enum CertificateType {
+  ACMPCA = 'acm',
+  FILE = 'file'
```
Trailing comma here:

```suggestion
  FILE = 'file',
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,110 @@
+  /**
+   * Contains information for your private certificate authority
+   */
+  readonly certificateAuthorityArns: acmpca.ICertificateAuthority[];
```
You missed a comment from my previous review: https://github.com/aws/aws-cdk/pull/11563#discussion_r532950230

```suggestion
  readonly certificateAuthorities: acmpca.ICertificateAuthority[];
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -165,14 +172,14 @@ export class VirtualGateway extends VirtualGatewayBase {
     } else {
       props.listeners.forEach(listener => this.listeners.push(listener.bind(this)));
     }
-
```
I would leave this as-is.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -60,6 +60,13 @@ export interface VirtualGatewayBaseProps {
+   *
+   * @default - No Config
+   */
+  readonly backendDefaultsClientPolicy?: ClientPolicy;
```
I think this name is slightly incorrect. There are multiple 'backends', but a single 'default'. So I would call this `backendsDefaultClientPolicy`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -94,6 +95,13 @@ export interface VirtualNodeBaseProps {
+   *
+   * @default - No Config
+   */
+  readonly backendDefaultsClientPolicy?: ClientPolicy;
```
Same comment here as for `VirtualGateway`.

```suggestion
  readonly backendsDefaultClientPolicy?: ClientPolicy;
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -187,14 +195,16 @@ export class VirtualNode extends VirtualNodeBase {
 
     props.backends?.forEach(backend => this.addBackend(backend));
     props.listeners?.forEach(listener => this.addListener(listener));
+
```
I would leave this as-is.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:
```diff
@@ -238,6 +238,48 @@ export = {
+          },
+        },
+      }));
+      test.done();
```
I usually like to leave an empty line before the last `test.done()`, makes the tests read slightly better in my view.

skinny85 reviewed (changes requested):
We're almost there! A few minor changes before we merge this in, mainly concerning naming and documentation.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -0,0 +1,110 @@
+  public bind(_scope: cdk.Construct): ClientPolicyConfig {
+    if (this.certificateType === CertificateType.ACMPCA && this.certificateAuthorityArns?.map(certificateArn =>
+      certificateArn.certificateAuthorityArn).length === 0) {
+      throw new Error('Certificate Authority ARN is required but empty');
```
Let's make this a little more clearer what and where was the problem:

```suggestion
      throw new Error('You must provide at least one Certificate Authority when creating an ACM Trust ClientPolicy');
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8b707cee7a22cf887b24dcdfd29e99a90e959141
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=47637410-db2a-41fa-9198-fab45fd5369d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (approved):
Looks great @sshver , thanks! 😊

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: de939c397be448dd17d7f1c6ff83a6cb08384fbe
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2a88e24b-9044-4207-af5f-1930002238d1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
